Add Large Forward Open service

nefethael opened the pull request:
Signed-off-by: Vincent Prince <vincent.prince.fr@gmail.com>

nefethael commented:
Hi all, 

This is my first contribution to EIPScanner, I'm not sure my PR takes care of everything (examples, doc, etc)?

Also, I only tried LFO feature with [OpENer](https://github.com/EIPStackGroup/OpENer/pull/303) stack and wireshark (I don't own any CIP adapter equipment), so I'm not 100% confident about implementation.

First commit allows output assembly to change data without vector copy, maybe a better solution would be to provide something like this in IOConnection? Current solution is not threadsafe.

```
using SendDataHandle = std::function<std::vector<uint8_t>&(void)>;
```

Second commit allows creation of ConnectionManager object as another class member without causing a timeout on first forward open.

Last commit adds large forward open service and provides a new NetworkConnectionParametersBuilder class that takes care of serializing/deserializing network connection parameters. It can be used this way

```
        ConnectionManager connectionManager;

        ConnectionParameters parameters;
        parameters.connectionPath = {0x20, 0x04,0x24, 151, 0x2C, 150, 0x2C, 100};
        parameters.o2tRealTimeFormat = true;
        parameters.originatorVendorId = 342;
        parameters.originatorSerialNumber = 32423;

        parameters.t2oNetworkConnectionParams =
                NetworkConnectionParametersBuilder(0, true)
                .SetConnectionType(NetworkConnectionParametersBuilder::P2P)
                .SetPriority(NetworkConnectionParametersBuilder::SCHEDULED)
                .SetConnectionSize(inputAssemblySize).Build();

        parameters.o2tNetworkConnectionParams =
                NetworkConnectionParametersBuilder(0, true)
                .SetConnectionType(NetworkConnectionParametersBuilder::P2P)
                .SetPriority(NetworkConnectionParametersBuilder::SCHEDULED)
                .SetConnectionSize(outputAssemblySize).Build();

        parameters.originatorSerialNumber = 0x12345;
        parameters.o2tRPI = 1000000;
        parameters.t2oRPI = 1000000;
        parameters.transportTypeTrigger |= NetworkConnectionParams::CLASS1;

        auto io = connectionManager.largeForwardOpen(si, parameters);
        ...
```

I'm open to any remarks & suggestions

Best regards,
Vincent

flipback commented:
Hey @nefethael , thank you for the contribution. 

We appreciate your help and we are open to work here together. Unfortunately, I'm pretty busy at the moment and cannot make the code review in details. I'll give you my feedback at the weekend.

Best regards, 
Aleksey

flipback commented on src/IOConnection.cpp:
```diff
@@ -40,14 +40,15 @@ namespace eipScanner {
 	IOConnection::~IOConnection() = default;
 
-	void IOConnection::setDataToSend(const std::vector<uint8_t> &data) {
+	void IOConnection::setDataToSend(std::vector<uint8_t> *data) {
```
Here you have broken the compatibility and the users might get an unexpected compilation error. Method setDataToSend received a const reference and no additional copy happened. Please, return the previous version.

nefethael replied:
How do you modify output assembly during processing? 
Are you calling setDataToSend every time you need to refresh the vector with full copy?

No problem to remove this commit from this PR as it's independant.

flipback replied:
Oh.. now I see what you mean. Yes, we copy the data into the internal buffer each time when it changes. For big vectors it is inefficient and your solution is better. However, I don't want to break the compatibility. 

I think we can make some workaround providing getter without 'const':

```cpp
std::vector<uint8_t>& IOConnection::getDataToSend() { return _outputData; }
```
A user can use it to mutate data without copping. It is a bit dirty way in may case, but it will work.

flipback commented on src/IOConnection.cpp:
```diff
@@ -85,15 +86,22 @@ namespace eipScanner {
+		auto now = std::chrono::steady_clock::now();
+		auto sinceLastHandle =
+			std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastHandleTime);
+			Logger(LogLevel::TRACE) << "Last call was " << sinceLastHandle.count() << "ms ago";
```
Unfortunately, there is an issue in our logger... This statement are calculated even if a user use a higher logging level. It could cause performance problems if you have a lot of IO connections.

nefethael replied:
Ok, I can remove Logger call then?

flipback replied:
Yes, please.

flipback commented on src/IOConnection.cpp:
```diff
@@ -85,15 +86,22 @@ namespace eipScanner {
 
-	bool IOConnection::notifyTick(std::chrono::milliseconds period) {
-		auto periodInMicroS = period.count() * 1000;
+	bool IOConnection::notifyTick() {
```
This change is fine for me, but I'd like t know your motivation?

nefethael replied:
I have a class that contains an instance of ConnectionManager. It is created at program launch, so when I execute ConnectionManager::handleConnections() method, saved _lastHandleTime is outdated regarding current time, causing new forward open connections to be immediatly be closed. 

Maybe another way is to let _lastHandleTime un-initialized and test for its validity at first connection? And set it back un-initialized when forward close is called?

flipback replied:
No, let's use your solution. I wanted to spare some time not to calculate the period between two calls of `notifyTick` in each connection. However, I think your solution is more robust and clean.

flipback commented on src/cip/connectionManager/NetworkConnectionParametersBuilder.h:
```diff
@@ -0,0 +1,64 @@
+namespace cip {
+namespace connectionManager {
+
+	class NetworkConnectionParametersBuilder {
```
This a great idea to make a builder for the connections parameters! One notice, our names of the methods starts from the low case: SetRedundantOwner -> setRedundantOwner

nefethael replied:
oops, I'm mixing our coding conventions with yours, I'll correct it.

nefethael commented:
@flipback I rebased PR branch with corrections. I just reverted `setDataToSend` stuff without any workaround for 0-copy vector, they're not that big regarding send period :)

flipback commented:
Hi @nefethael, thanks I see. What about other two topics?:
1) LOG_TRACE
2) Code conversion

nefethael commented:
@flipback I removed LOG_TRACE in notify function and fixed upper/lower-case issues, normally it should be ok

flipback commented:
@nefethael I didn't notice it because I was waiting for an additional commit =) Okay, I'll take  your PR. Thank you very much for your contribution again!
